nexus/db-model/src/affinity.rs

```diff
@@ -0,0 +1,85 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at https://mozilla.org/MPL/5.0/.
+
+// Copyright 2024 Oxide Computer Company
+
+//! Database representation of affinity and anti-affinity groups
+
+use super::impl_enum_type;
+use crate::schema::affinity_group;
+use crate::schema::affinity_group_instance_membership;
+use crate::schema::anti_affinity_group;
+use crate::schema::anti_affinity_group_instance_membership;
+use crate::typed_uuid::DbTypedUuid;
+use chrono::{DateTime, Utc};
+use omicron_uuid_kinds::AffinityGroupKind;
+use omicron_uuid_kinds::AntiAffinityGroupKind;
+use omicron_uuid_kinds::InstanceKind;
+
+impl_enum_type!(
+    #[derive(SqlType, Debug)]
+    #[diesel(postgres_type(name = "affinity_policy", schema = "public"))]
+    pub struct AffinityPolicyEnum;
+
+    #[derive(Clone, Copy, Debug, AsExpression, FromSqlRow, PartialEq)]
+    #[diesel(sql_type = AffinityPolicyEnum)]
+    pub enum AffinityPolicy;
+
+    // Enum values
+    Fail => b"fail"
+    Allow => b"allow"
+);
+
+impl_enum_type!(
+    #[derive(SqlType, Debug)]
+    #[diesel(postgres_type(name = "affinity_distance", schema = "public"))]
+    pub struct AffinityDistanceEnum;
+
+    #[derive(Clone, Copy, Debug, AsExpression, FromSqlRow, PartialEq)]
+    #[diesel(sql_type = AffinityDistanceEnum)]
+    pub enum AffinityDistance;
+
+    // Enum values
+    Sled => b"sled"
+);
+
+#[derive(Queryable, Insertable, Clone, Debug, Selectable)]
+#[diesel(table_name = affinity_group)]
+pub struct AffinityGroup {
+    pub id: DbTypedUuid<AffinityGroupKind>,
+    pub name: String,
+    pub description: String,
+    pub time_created: DateTime<Utc>,
+    pub time_modified: DateTime<Utc>,
+    pub time_deleted: Option<DateTime<Utc>>,
+    pub policy: AffinityPolicy,
+    pub distance: AffinityDistance,
+}
+
+#[derive(Queryable, Insertable, Clone, Debug, Selectable)]
+#[diesel(table_name = anti_affinity_group)]
+pub struct AntiAffinityGroup {
+    pub id: DbTypedUuid<AntiAffinityGroupKind>,
+    pub name: String,
+    pub description: String,
+    pub time_created: DateTime<Utc>,
+    pub time_modified: DateTime<Utc>,
+    pub time_deleted: Option<DateTime<Utc>>,
+    pub policy: AffinityPolicy,
+    pub distance: AffinityDistance,
+}
```

Comment:
Part of me wonders whether it makes sense to use the same model to represent these, since they share all the same fields...but I think the type-level distinction between what's an affinity group and what's an anti-affinity group will probably prevent bugs in the future, and using different typed UUIDs for the IDs is definitely the right move. Potentially shared bits like the policy and distance fields could be factored out later if there was a need to have functions that can operate on either, but at present it seems like there probably isn't!

Reply:
Yeah, I think it makes sense that they'll be able to evolve independently over time - I think the composition of group memberships described in RFD 522 inclined me to treat them separately.

(E.g., the idea that "anti-affinity groups can consist of instances or affinity groups" felt distinct enough from affinity groups, which, so far only consist of instances)

Reply:
Agreed! Carry on!